storagenode/piecestore/verification.go

```diff
@@ -71,7 +74,7 @@ func (endpoint *Endpoint) VerifyOrderLimit(ctx context.Context, limit *pb.OrderL
 	}
 
 	if err := endpoint.usedSerials.Add(ctx, limit.SatelliteId, limit.SerialNumber, serialExpiration); err != nil {
-		return ErrVerifyDuplicateRequest.Wrap(err)
+		return status.Errorf(codes.AlreadyExists, "serial number is already used: %+v", err)
```

Comment:
Wondering whether `codes.Unauthenticated` or something else would be more appropriate here.

Reply:
For what I understood that the `usedSerials.Add` this error made sense to me, but I'm not fully sure what serials do.

Reply:
Serials are for preventing replay attacks... i.e. such that a single person cannot submit the same order limit twice.

Reply:
Then it doesn't look to me that `codes.Unauthenticated` fits better. Perhaps `codes.FailedPrecondition` may be better, but this one doesn't look bad to me.

Reply:
FailedPrecondition are about the state of the target object. I.e. that there might be a call that would fix the state and allow the request to succeed.

Unauthenticated means that the authentication is not valid. Nonce and signature are part of the message authentication.

InvalidArgument might also be applicable.

https://github.com/grpc/grpc/blob/master/doc/statuscodes.md.

Reply:
PermissionDenied could also be used.

Reply:
https://oauth.net/core/1.0a/#http_codes

Reply:
I agree about `FailedPrecondition` sorry I didn't remember well the description and I didn't check again before commenting.

`InvalidArgument` sounds good but I don't see that the current one is wrong anyway.

Reply:
`codes.AlreadyExists` is about the resource itself... e.g. trying to upload the same piece id twice. I would go with `Unauthenticated` since it's seems it's used for things related to credentials failure.

Reply:
I'm still not getting why is `Unauhhtenticated`, the client should be fine the only thing that's doing wrong is sending an order limit which has already been used.

I would go with `InvalidArgument`

Reply:
Because a bad nonce is usually treated as part of the credentials... it's similar to having a wrong signature or bad signer.

Reply:
I'm still not understanding what the orders stuff, so I cannot confirm that.

I'll change to `Unauthenticated`